In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import fetch_california_housing
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error

%matplotlib inline
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 100)

print("✅ Библиотеки загружены!")
print(f"pandas: {pd.__version__}")
print(f"numpy: {np.__version__}")

✅ Библиотеки загружены!
pandas: 1.4.2
numpy: 1.21.5


In [2]:
# Загружаем данные
housing = pd.read_csv(open('housing.csv'))

# Создаем DataFrame
df = housing

print("Размер данных:", df.shape, "строки, столбцы")
print("\nПервые 5 строк:")
display(df.head())
print("\nИнформация о столбцах:")
display(df.info())
print("\nОписательные статистики:")
display(df.describe())
print("\nПроверка на пропуски:")
df.isnull().sum()

Размер данных: (20640, 10) строки, столбцы

Первые 5 строк:


,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,-122.23,37.88,41.0,880.0,129.0,322.0,126.0,8.3252,452600.0,NEAR BAY
1,-122.22,37.86,21.0,7099.0,1106.0,2401.0,1138.0,8.3014,358500.0,NEAR BAY
2,-122.24,37.85,52.0,1467.0,190.0,496.0,177.0,7.2574,352100.0,NEAR BAY
3,-122.25,37.85,52.0,1274.0,235.0,558.0,219.0,5.6431,341300.0,NEAR BAY
4,-122.25,37.85,52.0,1627.0,280.0,565.0,259.0,3.8462,342200.0,NEAR BAY



Информация о столбцах:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


None


Описательные статистики:


,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value
count,20640.000000,20640.000000,20640.000000,20640.000000,20433.000000,20640.000000,20640.000000,20640.000000,20640.000000
mean,-119.569704,35.631861,28.639486,2635.763081,537.870553,1425.476744,499.539680,3.870671,206855.816909
std,2.003532,2.135952,12.585558,2181.615252,421.385070,1132.462122,382.329753,1.899822,115395.615874
min,-124.350000,32.540000,1.000000,2.000000,1.000000,3.000000,1.000000,0.499900,14999.000000
25%,-121.800000,33.930000,18.000000,1447.750000,296.000000,787.000000,280.000000,2.563400,119600.000000
50%,-118.490000,34.260000,29.000000,2127.000000,435.000000,1166.000000,409.000000,3.534800,179700.000000
75%,-118.010000,37.710000,37.000000,3148.000000,647.000000,1725.000000,605.000000,4.743250,264725.000000
max,-114.310000,41.950000,52.000000,39320.000000,6445.000000,35682.000000,6082.000000,15.000100,500001.000000



Проверка на пропуски:


longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [3]:
median_bedrooms = df['total_bedrooms'].median()
df['total_bedrooms'] = df['total_bedrooms'].fillna(median_bedrooms)

print("Пропуски в total_bedrooms после заполнения:", df['total_bedrooms'].isnull().sum())

# Целевая переменная
y = df['median_house_value']

# Матрица признаков (удаляем целевую переменную)
X = df.drop('median_house_value', axis=1)

print("Размер X:", X.shape)
print("Размер y:", y.shape)
print("\nИмена признаков (столбцы X):", X.columns.tolist())

# Еще раз убедимся, что в X нет пропусков и все числовое
print("\nПропуски в X:", X.isnull().sum().sum())
print("Типы данных в X:\n", X.dtypes)

Пропуски в total_bedrooms после заполнения: 0
Размер X: (20640, 9)
Размер y: (20640,)

Имена признаков (столбцы X): ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'ocean_proximity']

Пропуски в X: 0
Типы данных в X:
 longitude             float64
latitude              float64
housing_median_age    float64
total_rooms           float64
total_bedrooms        float64
population            float64
households            float64
median_income         float64
ocean_proximity        object
dtype: object


In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Размер обучающей выборки (X_train):", X_train.shape)
print("Размер тестовой выборки (X_test):", X_test.shape)
print("Размер целевой переменной на обучении (y_train):", y_train.shape)
print("Размер целевой переменной на тесте (y_test):", y_test.shape)

Размер обучающей выборки (X_train): (16512, 9)
Размер тестовой выборки (X_test): (4128, 9)
Размер целевой переменной на обучении (y_train): (16512,)
Размер целевой переменной на тесте (y_test): (4128,)


In [5]:
from sklearn.linear_model import LinearRegression

# Преобразуем категориальный признак в "фиктивные" переменные
X_train_encoded = pd.get_dummies(X_train, columns=['ocean_proximity'], drop_first=True)
X_test_encoded = pd.get_dummies(X_test, columns=['ocean_proximity'], drop_first=True)

# Синхронизируем столбцы: добавляем отсутствующие в test и удаляем лишние
X_test_encoded = X_test_encoded.reindex(columns=X_train_encoded.columns, fill_value=0)

# Убеждаемся, что количество столбцов совпадает
print(f"Столбцов в X_train: {X_train_encoded.shape[1]}")
print(f"Столбцов в X_test: {X_test_encoded.shape[1]}")

# Обучение модели
model = LinearRegression()
model.fit(X_train_encoded, y_train)

print("Модель обучена! ✅")

# Предсказания
y_pred = model.predict(X_test_encoded)

print("Предсказания сделаны! ✅")

Столбцов в X_train: 12
Столбцов в X_test: 12
Модель обучена! ✅
Предсказания сделаны! ✅


In [6]:
# Создаем DataFrame для сравнения (первые 10 значений)
comparison_df = pd.DataFrame({
    'Реальное значение': y_test[:10],
    'Предсказанное значение': y_pred[:10],
    'Разница': y_test[:10] - y_pred[:10]
})

print("Сравнение предсказаний с реальными значениями (первые 10):")
display(comparison_df)

Сравнение предсказаний с реальными значениями (первые 10):


,Реальное значение,Предсказанное значение,Разница
20046,47700.0,54055.448899,-6355.448899
3024,45800.0,124225.338937,-78425.338937
15663,500001.0,255489.379492,244511.620508
20484,218600.0,268002.431569,-49402.431569
9814,278000.0,262769.434816,15230.565184
13311,158700.0,139606.303956,19093.696044
7113,198200.0,290665.423914,-92465.423914
7668,157500.0,228264.876375,-70764.876375
18246,340000.0,256506.785610,83493.214390
5723,446600.0,407923.858435,38676.141565


In [7]:
mae = mean_absolute_error(y_test, y_pred)
rmse = np.sqrt(mean_squared_error(y_test, y_pred))

print(f"MAE на тестовой выборке: {mae:.4f}", '                             (средняя абсолютная ошибка)')
print(f"RMSE на тестовой выборке: {rmse:.4f}", '                            (среднеквадратичная ошибка)')

# Сравнение со средним значением y
mean_y = y.mean()
print(f"Среднее значение целевой переменной (median_house_value): {mean_y:.4f}")
print(f"Отношение MAE к среднему: {mae / mean_y:.2%}")
print(f"Отношение RMSE к среднему: {rmse / mean_y:.2%}")

MAE на тестовой выборке: 50670.7382                              (средняя абсолютная ошибка)
RMSE на тестовой выборке: 70060.5218                             (среднеквадратичная ошибка)
Среднее значение целевой переменной (median_house_value): 206855.8169
Отношение MAE к среднему: 24.50%
Отношение RMSE к среднему: 33.87%


- MAE (Mean Absolute Error) показывает среднюю абсолютную ошибку модели в тех же единицах, что и целевая переменная. В нашем случае, модель ошибается в среднем на 50.7 тысяч долларов.

- RMSE (Root Mean Squared Error) аналогичен, но сильнее штрафует за большие ошибки, так как возводит их в квадрат. В нашем случае RMSE = 70060 долларов.

Чтобы понять, велика ли ошибка, сравниваем ее со средним значением y (около 2.07). MAE составляет около 24.5%, что говорит о том, что ошибка довольно значительна, и модель не является идеальной, но дает некоторое приближение. RMSE составляет около 33.87% от среднего.

In [8]:
y_test_pred = model.predict(X_test_encoded)
y_train_pred = model.predict(X_train_encoded)

mae_train = mean_absolute_error(y_train, y_train_pred)
rmse_train = np.sqrt(mean_squared_error(y_train, y_train_pred))

print(f"MAE на обучающей выборке: {mae_train:.4f}")
print(f"MAE на тестовой выборке: {mae:.4f}")
print(f"Разница MAE (Train - Test): {mae_train - mae:.4f}")

MAE на обучающей выборке: 49594.8421
MAE на тестовой выборке: 50670.7382
Разница MAE (Train - Test): -1075.8961


Переобучение (overfitting) — это ситуация, когда модель хорошо работает на обучающих данных, но значительно хуже на новых (тестовых). В нашем случае разница небольшая, поэтому можно сделать вывод, что сильного переобучения нет. Модель обобщается достаточно хорошо.

In [9]:
# Веса модели (коэффициенты)
weights = model.coef_

# Получаем названия признаков после one-hot encoding
feature_names = X_train_encoded.columns.tolist()

# Создаем DataFrame для наглядности
feature_importance = pd.DataFrame({
    'Признак': feature_names,
    'Вес (коэффициент)': weights
})

print("Вес признаков в модели:")
display(feature_importance.sort_values(by='Вес (коэффициент)', ascending=False))

Вес признаков в модели:


,Признак,Вес (коэффициент)
9,ocean_proximity_ISLAND,136125.072615
7,median_income,39473.975175
11,ocean_proximity_NEAR OCEAN,3431.140073
2,housing_median_age,1102.185084
4,total_bedrooms,102.789395
6,households,48.252753
3,total_rooms,-6.021506
5,population,-38.172906
10,ocean_proximity_NEAR BAY,-5136.642217
1,latitude,-25468.352050


- Модель линейной регрессии использует формулу y = X·w, где w — это веса (коэффициенты). Знак веса показывает направление влияния: положительный — признак увеличивает цену, отрицательный — уменьшает. 
Модуль веса показывает силу влияния (при прочих равных).

##Самые сильные положительные факторы (увеличивают стоимость):

1. **ocean_proximity_ISLAND** (+136 125) — дома на островах значительно дороже. Это логично: островная недвижимость редкая и престижная.

2. **median_income** (+39 474) — чем выше доход в районе, тем дороже жильё. Это полностью соответствует реальной жизни: в богатых районах цены выше.

3. **ocean_proximity_NEAR OCEAN** (+3 431) — близость к океану повышает стоимость жилья.

##Самые сильные отрицательные факторы (уменьшают стоимость):

1. **ocean_proximity_INLAND** (-39 787) — удалённость от океана **сильнее всего снижает** стоимость жилья. Это логично: люди готовы платить больше за близость к воде.

2. **longitude** (-26 838) и **latitude** (-25 468) — географическое положение влияет на цену: в некоторых регионах (например, внутренних районах) цены ниже.

3. **ocean_proximity_NEAR BAY** (-5 137) — близость к заливу тоже снижает цену (возможно, из-за менее престижного расположения).

###Вывод:

Модель выделила **доход** и **близость к океану** как ключевые факторы, что совпадает с житейской интуицией. 
- Жильё на островах и у океана дороже.
- Внутренние районы и удалённые локации — дешевле.
- Доход района — главный числовой фактор.

Веса — это коэффициенты линейной комбинации **y = X·w**, которые модель подобрала автоматически на основе обучающих данных.
Такой анализ помогает не только предсказывать, но и интерпретировать модель, что очень важно для бизнеса.